README.md

```diff
@@ -2048,6 +2049,214 @@ config.SetProperty(property.StatisticsPeriodSeconds.Name(), "4")
+
+```
+[Time] [caller method name]
+[Log level] [Client Name] [Group Name] [Client Version] [Log Message]
+```
+
+An example default log message is as follows:
+```
+2018/11/30 17:48:52 github.com/hazelcast/hazelcast-go-client/internal.(*lifecycleService).fireLifecycleEvent
+INFO:  hz.client_1 [dev] [0.4] New State :  STARTED
+```
+
+If you want to modify the Default Logger for your convenience, you can do so by accessing the embedded built-in go logger:
+```go
+l := logger.New()
+l.SetPrefix("myPrefix ")
+config := hazelcast.NewConfig()
+config.LoggerConfig().SetLogger(l)
+``` 
+
+The same log message will now be as follows:
+
+```
+myPrefix 2018/11/30 17:55:40 github.com/hazelcast/hazelcast-go-client/internal.(*lifecycleService).fireLifecycleEvent
+INFO:  hz.client_1 [dev] [0.4] New State :  CONNECTED
+```
+
+
+If you want to set a custom logger, you can do so by implementing `Logger` interface:
+
+```go
+type customLogger struct {
+}
+
+func (c *customLogger) Debug(args ...interface{}) {
+}
+
+func (c *customLogger) Trace(args ...interface{}) {
+}
+
+func (c *customLogger) Info(args ...interface{}) {
+	log.Println(args)
+}
+
+func (c *customLogger) Warn(args ...interface{}) {
+}
+
+func (c *customLogger) Error(args ...interface{}) {
+}
+```
+
+Note that the `customLogger` only logs `Info` level to console. This way users can
+implement only the levels they need in the format they want to.
+
+After implementing the `Logger` interface, you need to set it as the client's logger:
+
+```go
+customLogger := &customLogger{}
+config := hazelcast.NewConfig()
+config.LoggerConfig().SetLogger(customLogger)
+```
+
+You can also integrate any third party logger with `Logger` interface:
+
+#### Logrus
+
+```go
+logger := logrus.New()
+logger.SetLevel(logrus.DebugLevel)
+config := hazelcast.NewConfig()
+config.LoggerConfig().SetLogger(logger)
+```
+
+The same log message will now be:
+
+```
+time="2018-11-30T18:09:52+03:00" level=info msg="New State : CONNECTED"
+```
+
+Note that Logrus already implements the `Logger` interface, therefore we did not need to do any extra work.
+
+
+#### Zap
+
+```go
+type zapLogger struct {
+	*zap.Logger
+}
+
+func (z *zapLogger) Debug(args ...interface{}) {
+	message := fmt.Sprintln(args)
+	z.Logger.Debug(message)
+}
+
+func (z *zapLogger) Trace(args ...interface{}) {
+	// NO OP
+}
+
+func (z *zapLogger) Info(args ...interface{}) {
+	message := fmt.Sprintln(args)
+	z.Logger.Info(message)
+}
+
+func (z *zapLogger) Warn(args ...interface{}) {
+	message := fmt.Sprintln(args)
+	z.Logger.Warn(message)
+}
+
+func (z *zapLogger) Error(args ...interface{}) {
+	message := fmt.Sprintln(args)
+	z.Logger.Error(message)
+}
+```
+
+Then, you need to set it as the client's logger:
+
+```go
+opt := zap.AddCallerSkip(1)
+logger, _ := zap.NewProduction(opt)
+zapLogger := &zapLogger{logger}
+config := hazelcast.NewConfig()
+config.LoggerConfig().SetLogger(zapLogger)
+```
+
+Note that we call `zap.AddCallerSkip(1)` to skip our wrapper method.
```

Comment:
can you explain the reason for this?

Reply:
Let's say we have a method x:

```
func x()  {
    logger.Info(..)
}
```

If we expand `logger.Info()` in `x()`, we have:

```
func (z *zapLogger) Info(args ...interface{}) {
 	message := fmt.Sprintln(args)
 	z.Logger.Info(message)
 }
```

So `z.Logger.Info(message)` will be called from our wrapper method, which means that the caller of `z.Logger.Info(message)` is: `func (z *zapLogger) Info(args ...interface{})`. However we want it to be `x`, so we will skip one caller, which is `func (z *zapLogger) Info(args ...interface{})`, and the caller will be `x`.